### DISCUSSION TITLE
[proposal] Remove -Werror compiler flag from AGW c/c++ build targets

### DISCUSSION THREAD

**Discussion by electronjoe** [magma/magma#5273] [Ideas]
# Proposal Summary

Turn off -Werror compiler flags across c/c++ build targets (only exist today on AGW - see [List](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/magma/magma%24%4059e1695+-Werror&patternType=literal)).

## Benefits of Werror

I can think of two benefits of Werror use in some settings and teams:

- It improves the visibility of newly introduced Warnings for the developer
- In the event a team suffers from contributors knowingly committing code with warnings, it prevents commit

## Drawbacks of Werror

Drawbacks of Werror include:

- Teams with Werror turned on are more prone to pruning warning flags over time
  - Magma codebase is broadly running light on warning flags (maybe why)
- Turning on new warnings is harder
  - All existing warnings must be dealt with before the warnings can be made visible to the team
- Migration to new tooling requires more effort
  - New tooling invariably has better static analysis / warnings
  - Werror use prevents new tooling from being migrated until all new warnings are dealt with

## Net Benefit

I believe that by tracking Warning counts in CI (perhaps by warning type even) - see #4866 - we achieve the Visibility objective that Werror is providing us.  This leaves only the second benefit of blocking contributors knowingly introducing new warnings.  I would like to see that this is a problem before introducing this overhead.

## Additional Background

My experience is largely aligned with that of this blog post: [Werror is not your friend](https://embeddedartistry.com/blog/2017/05/22/werror-is-not-your-friend/).

Or the following [Gnome Desktop Manager PR Comments](https://mail.gnome.org/archives/desktop-devel-list/2013-February/msg00098.html):

> I'll be the first to agree that -Werror is evil stuff, we even fell into this catch 22 only around a year ago where one group thought it was a good idea to infect our builds with -Werror, while another group thought it was a good idea to add deprecation warnings, ... I'm not even sure we're out of the woods from that tangled mess yet. (and I'll side with the latter, deprecation warnings are a good thing, warnings are a good thing in general).

Or this post on Rust: [PSA: `deny(warnings)` is actively harmful](https://www.reddit.com/r/rust/comments/f5xpib/psa_denywarnings_is_actively_harmful/)

**Comment by electronjoe** [comment]
I have had several experiences with Werror problems already at Magma - around tooling upgrades and inconsistent CMake configuration / dependency behavior.  For example, within lte/gateway - in my new Ubuntu 20.04 docker container, `test_oai` successfully compiles  `lte/protos/mconfig.proto` while `test_session_manager` target fails to build due `lte/protos/mconfig.proto` as for some reason this target applies `-Werror` to the proto builds, while the former does not. Further, the warning is over use of deprecated Enums - which is by design in the output of protoc (one must reference the DEPRECATED enums in the proto outputs, clealry).  Tracing this over an hour yesterday I have not yet figured out why Werror is applied to one target and not the other. This is a CMake issue - and it should be fixed, but it should not be causing blockage to build tool improvements IMO.

As a second example, see the size of the PR required to turn on UBSAN (PR #5262).  This is pretty complex code and ideally would be chopped into a number of PRs. I suppose nothing prevents you from doing that, submitting the PRs, then finally committing the flag update - but that is setting a high bar for very busy developers, to complete all fixups before turning on potentially critical warnings.

**Comment by themarwhal** [comment]
Looks like gcc9 has a way to output the warning into JSON: https://developers.redhat.com/blog/2019/03/08/usability-improvements-in-gcc-9/
Maybe we can upload this to CircleCI to track the stats.

**Reply by electronjoe** [reply]
I dug into CircleCI and was unable to find any support for tracking custom stats.

So I have draft PRs up to post Clang warnings and Clang-Tidy finding counts to a Google Doc - as at least an initial prototype. Kinda a bummer to have to turn up yet another third party service - a Gdoc seemed light weight.

I also looked for tooling to parse / post this in a more structured way and came up empty - hence my present prototype =/
